File: services/rust/actix/src/routes/handlers/patch.rs
use crate::routes::{
    errors::AppError,
    extractors::{AppJson, AppPath},
};
use actix_web::{HttpResponse, ResponseError, web::Data};
use app::{state::AppState, task::PatchedTask};
use std::sync::Mutex;
use tracing::{error, info, instrument, warn};

#[instrument(skip_all)]
pub async fn patch_handler(
    state: Data<Mutex<AppState>>,
    AppPath(id): AppPath,
    AppJson(new_task): AppJson<PatchedTask>,
) -> HttpResponse {
    if let Ok(mut state) = state.lock() {
        if let Some(task) = state.tasks.get_mut(&id) {
            let request = new_task;

            let previous_operation = task.operation.clone();
            task.operation = request.operation;

            info!(
                %id,
                secret = task.secret.len(),
                from_operation = previous_operation.to_string(),
                to_operation = task.operation.to_string(),
                method = "PATCH",
                "Patched task"
            );

            return HttpResponse::Ok().json(task);
        } else {
            drop(state);
            warn!(%id, method = "PATCH", "Task not found");
            return AppError::TaskNotFound.error_response();
        }
    }

    error!(%id, method = "PATCH", "Poisoned lock");

    AppError::Internal.error_response()
}

File: services/go/gin/routes/handlers/patch.go
package handlers

import (
	"app"
	"gin/routes"
	"log/slog"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
)

func PatchHandler(state *app.AppState) gin.HandlerFunc {
	return func(c *gin.Context) {
		id, err := ParseID(c)
		if err != nil {
			return
		}

		body, err := ReadBody(c)
		if err != nil {
			return
		}

		patchedTask, err := ParsePatchedTask(c, body)
		if err != nil {
			return
		}

		state.Mu.Lock()
		task, ok := state.Tasks[id]

		if !ok {
			state.Mu.Unlock()
			slog.Warn("Task not found", "method", c.Request.Method, "path", c.Request.URL.Path, "id", id.String())
			routes.AppErrors.TaskNotFound.Error(c)
			return
		}

		previousOperation := task.Operation
		task.Operation = patchedTask.Operation
		state.Tasks[id] = task
		state.Mu.Unlock()

		slog.Info(
			"Patched task",
			"id", id.String(),
			"secret", len(task.Secret),
			"from_operation", strings.ToLower(string(previousOperation)),
			"to_operation", strings.ToLower(string(patchedTask.Operation)),
			"method", c.Request.Method,
		)

		c.JSON(http.StatusOK, task)
	}
}
